### ISSUE TITLE
ADFS: relying party identifier configuration

### ISSUE THREAD

**Issue by lcgkm** [vouch/vouch-proxy#189]
I found vouch-proxy use RedirectURL as relying party identifier by default.
Could we configure the relying party identifier in config.yml by ourselves?
```
func setDefaultsADFS() {
	log.Info("configuring ADFS OAuth")
	OAuthopts = oauth2.SetAuthURLParam("resource", GenOAuth.RedirectURL) // Needed or all claims won't be included
}
```

**Comment by bnfinet** [comment]
for context..

URIs as object identifiers: relying party identifier
https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/how-uris-are-used-in-ad-fs
 
If the RPI is a URI is it okay to have it be https://vouch.yourdomain.com/auth?

Is that a problem for you?

**Comment by lcgkm** [comment]
However, the relying party identifier is usually a UUID, not a URI.
![image](https://user-images.githubusercontent.com/5523451/71154651-f7f90980-2276-11ea-96c7-98cdc3dea075.png)

**Comment by lcgkm** [comment]
>Is that a problem for you?
RedirectURL is not equal relying party identifier. 
It's hard to talk with our ADFS admin. They don't allow us add RedirectURL as relying party identifier.

**Comment by bnfinet** [comment]
@simongottschlag what do you think of this suggestion?

**Comment by simongottschlag** [comment]
@bnfinet @lcgkm I haven't had the problem since we usually manage ADFS ourselves. I don't see it as a problem making it configurable - I just hadn't thought about it when implementing it.

Maybe make sure to use redirectUri if none is defined in the configuration, to keep backward compatibility?

**Comment by bnfinet** [comment]
@lcgkm I know this is an old thread but I wanted to clarify that I'd be open to a PR

**Comment by tangel0v** [comment]
Maybe I can open an PR to you with this @bnfinet as I have been stuck in my other issue and if I fix this, I'll fix my problem too.

**Comment by bnfinet** [comment]
@tangel0v are you saying #257 is related to this issue?

Thanks for the offer of a PR, that's always appreciated.  However...

> In the last versions of Microsoft ADFS you can pass the Relaying Party Trust as part of the scopes of OAuth (for example: "$MY_PRECIOUS_APP/openid allatclaims" as scopes will set openid and allatclaims as scopes and $MY_PRECIOUS_APP as Relaying Party Trust ID.

If ADFS sets the RPI to `$MY_PRECIOUS_APP` when `oauth.scopes` includes `$MY_PRECIOUS_APP/openid` would it be better to just document this?  Would that be sufficient to satisfy the feature request?  That'd be better than adding a new knob to Vouch Proxy IMHO.

**Comment by tangel0v** [comment]
Not exactly. I was facing this issue in my configuration therefore I tried to use the standard oidc.

Putting the Redirect URL as Relying Party Trust ID just works (in ADFS versions 2016 and 2019 at least), but it brokes all our RP IDs naming convention: that's the reason why I was trying to use oidc.

**Comment by bnfinet** [comment]
@tangel0v makes sense, a PR would be welcome

**Comment by mario-tux** [comment]
@tangel0v Could you explain in details if there is a way to make ADFS works with VP without PR but with a configuration of ADFS where the RPI is not the RedirectURL? We can't change the ADFS configuration (different management). It looks that patching the code of VP to not set the `resource` field (remarking the setting in `setDefaultsADFS` function), as in a oidc-like configuration, it just works.

@bnfinet If you are available we can provide a PR that, on request, can pass a specific RPI in a adfs configuration.

@Herbrant

**Comment by mario-tux** [comment]
We submitted a PR (https://github.com/vouch/vouch-proxy/pull/393) to cover the OP support request: it should guarantee backward compatibility. It works in our application scenario with a UUID as `relying_party_id`. Please, consider its mainstream inclusion.

@Herbrant
